File: 443-string-compression/string-compression.py
class Solution:
    def compress(self, chars: list[str]) -> int:
        write = 0  # pointer for writing compressed result
        i = 0      # pointer for reading

        while i < len(chars):
            char = chars[i]
            count = 1

            # Count consecutive occurrences of the current character
            while i + count < len(chars) and chars[i + count] == char:
                count += 1

            # Write the character
            chars[write] = char
            write += 1

            # Write the count if it's more than 1
            if count > 1:
                for digit in str(count):
                    chars[write] = digit
                    write += 1

            # Move to the next different character
            i += count

        return write

File: 443-string-compression/string-compression.cpp
class Solution {
public:
    int compress(vector<char>& chars) {
        int write = 0, i = 0;
        
        while (i < chars.size()) {
            char c = chars[i];
            int cnt = 0;
            while (i < chars.size() && chars[i] == c) {
                cnt++;
                i++;
            }
            
            chars[write++] = c;
            
            if (cnt > 1) {
                string s = to_string(cnt);
                for (char ch : s) {
                    chars[write++] = ch;
                }
            }
        }
        
        return write;
    }
};